### PR TITLE
Fix prvalue to xvalue conversion address space assignment

### PR DESCRIPTION
**Pull request by Naghasan**
The common address deduction rules assumed prvalue-to-xvalue conversion can only happens inside a function.
The patch aligns the rules to match declaration rules.

It is a bit of a corner case, but the spec treated all 3 conversions bellow in the same way.

```cpp
constexpr const int& foo = 5;

class Foo {
public:
  static constexpr const int& foo = 42;
};

void bar() {
    const int& foo = 42;
}
```

### CODE REVIEW

File: adoc/chapters/device_compiler.adoc [hunk 1]
```diff
@@ -748,9 +748,11 @@ code called by the device kernel.
   ** Static data members are treated the same way as for variable in
      namespace scope
 
-The result of a prvalue-to-xvalue conversion is assigned to the local
-address space if it happens in a hierarchical context or to the private
-address space otherwise.
+If a prvalue-to-xvalue conversion happens in a block scope or function parameter scope,
+the result is assigned to the local address space if it happens a hierarchical context
```
Comment (keryell) [inline adoc/chapters/device_compiler.adoc:752]:
```suggestion
the result is assigned to the local address space if it happens in a hierarchical context
```

**Comment by keryell** [comment]
@Naghasan ping!

File: adoc/chapters/device_compiler.adoc (hunk 1 above)
Comment (Naghasan) [reply adoc/chapters/device_compiler.adoc]: I addressed it in the follow up commit.

**Comment by Naghasan** [comment]
@keryell  I extended the initial text to also account for the context in which prvalue to xvalue is taking place. I find it a bit "hand wavy" but still acceptable.

The new commit also account for this:

```cpp
void bar() {
    static constexpr const int& foo = 42;
}
```

**Comment by Naghasan** [comment]
Resolved conflict: rebased to head and ran reflow

File: adoc/chapters/device_compiler.adoc [hunk 2]
```diff
@@ -737,8 +737,8 @@ The variable declarations get assigned to an address space depending on their
 scope and storage class:
 
   * Namespace scope
-  ** If the type is [code]#const#, the address space the declaration is assigned
-     to is implementation-defined.
+  ** If the type is [code]#const#, the address space of the declaration is
+     assigned to is implementation-defined.
```
Comment (gmlueck) [inline adoc/chapters/device_compiler.adoc:741]: This wording no longer makes sense.  The original wording was a little awkward, but it did make sense.  Are we trying to say this:

> If the type is `const`, the declaration is assigned to an implementation-defined address space.

File: adoc/chapters/device_compiler.adoc [hunk 3]
```diff
@@ -761,9 +761,14 @@ device kernel or code called by the device kernel.
   ** Static data members are treated the same way as for variable in namespace
      scope
 
-The result of a prvalue-to-xvalue conversion is assigned to the local address
-space if it happens in a hierarchical context or to the private address space
-otherwise.
+If a prvalue-to-xvalue conversion happens as part of the an initialization
```
Comment (gmlueck) [inline adoc/chapters/device_compiler.adoc:764]:
```suggestion
If a prvalue-to-xvalue conversion happens as part of an initialization
```

File: adoc/chapters/device_compiler.adoc (hunk 2 above)
Comment (Naghasan) [reply adoc/chapters/device_compiler.adoc:741]: reads better, I'll apply this.

**Review by gmlueck — APPROVED** [review]

**Pull request merged**